src/environment.rs

```diff
@@ -120,6 +101,7 @@ pub enum LockFileUsage {
 pub async fn get_up_to_date_prefix(
```

Comment:
I still had to re-read this function a couple of times to understand what is going on. I think there are a lot of diverging if branches/booleans that make it hard to read.

I was thinking that it might make sense to add a little bit of duplication by adding for instance two separate code paths for `no_install`/`do_install`. So have a function that is called (from this function) for the `no_install` case and one for when installation should be done.

Reply:
There is also the weird case where if you do a Frozen no_install the function essentially does nothing!

I'm fine with having two different functions that are re-used before this it was duplicated 11 times or something which seems a bit much.

Another option would be an option struct or something and matching on this and having a couple of different code paths for the options.